fix(core): fix removal of a container reference used in the component…

aparzi opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [X] Bugfix


## What is the current behavior?
ng generate @angular/core:control-flow migration removes the container reference even if it is used in component file

Issue Number: #56565

## What is the new behavior?

During migration a container reference was deleted even though it was used in the component file.

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

JeanMeche commented:
Could you please capture the fix in a unit test in `packages/core/schematics/test/control_flow_migration_spec.ts`. 
Thank you.

aparzi commented:
> Could you please capture the fix in a unit test in `packages/core/schematics/test/control_flow_migration_spec.ts`. You can run the tests with `yarn bazel test //packages/core/schematics/test/...`.
> 
> Thank you.

Hi @JeanMeche I Integrated unit test, based on the issue template

JeanMeche commented on packages/core/schematics/ng-generate/control-flow-migration/util.ts:
```diff
@@ -514,6 +514,12 @@ export function processNgTemplates(template: string): {migrated: string; err: Er
 }
 
+function checkViewChildTemplate(templateName: string, file: string): boolean {
+    const viewChildRegex = /@ViewChild\(['"`](.*?)['"`]/;
```
We probably need to cover ViewChildren too.

aparzi replied:
Ok @JeanMeche, I agree, thanks.

Do you prefer that I make a dedicated function for viewChildren or do you think it's better to extend the behavior of this one?

aparzi commented:
Hi @JeanMeche,
I updated the PR with dealing ViewChildren. I created a dedicated function so as to keep the management of the different decorators separate.

Thanks.

JeanMeche commented on packages/core/schematics/ng-generate/control-flow-migration/util.ts:
```diff
@@ -514,6 +521,24 @@ export function processNgTemplates(template: string): {migrated: string; err: Er
+}
+
+function checkViewChildTemplate(templateName: string, file: string): boolean {
+  const viewChildRegex = /@ViewChild\(['"`](.*?)['"`]/;
```
Ideally this shouldn't be done via a regex but by analysing the AST. This would prevent commented code to return a false positive for example.

aparzi replied:
ok I'll make the change right away.

Thanks

aparzi replied:
@JeanMeche updated analysis with AST and updated unit test .

JeanMeche commented on packages/core/schematics/ng-generate/control-flow-migration/util.ts:
```diff
@@ -514,6 +521,26 @@ export function processNgTemplates(template: string): {migrated: string; err: Er
+        ts.isIdentifier(expr.expression) &&
+        (expr.expression.text === 'ViewChild' || expr.expression.text === 'ViewChildren')
+      ) {
+        hasActiveViewChild = true;
```
Quick optim would be an early return here.

aparzi replied:
I inserted an exit condition so as to have an improvement in the iteration

JeanMeche reviewed (approved):
Beside the nit, LGTM. 

Would also like to have @thePunderWoman's approval on this!

aparzi commented:
> Beside the nit, LGTM.
> 
> Would also like to have @thePunderWoman's approval on this!

Ok @JeanMeche I'll make the improvement as soon as possible. Thanks

JeanMeche commented on packages/core/schematics/ng-generate/control-flow-migration/util.ts:
```diff
@@ -495,7 +498,11 @@ export function processNgTemplates(template: string): {migrated: string; err: Er
+        if (
+          t.count === matches.length + 1 &&
+          safeToRemove &&
+          !checkViewChildOrViewChildrenNodes(sourceFile)
```
Actually this test will keep any template if there is one ViewChild query, even if the query doesn't reference the template. 

Should probably pass the template name and check for it in the AST.

aparzi replied:
Do you want me to handle this scenario? Like if a viewChild has a query to a template that doesn't exist? 

Example:
```
@Component({
    selector: 'app-root',
    imports: [
        NgIf
    ],
    template: `
        <h1>Hello from {{ name }}!</h1>
        <div *ngIf="showContent; then contentTemplate"></div>
        <ng-template #contentTemplate>
            <div>test content</div>
        </ng-template>
        `
})
export class AppComponent implements AfterViewInit {
    @ViewChild('contentTemplatesWrong') testContainer!: TemplateRef<unknown>;
    name = 'Angular';
    showContent = true;

    constructor(private viewContainerRef: ViewContainerRef) {}

    ngAfterViewInit(): void {
        this.viewContainerRef.createEmbeddedView(this.testContainer);
    }
}
```

JeanMeche replied:
The query `ViewChild(MyDirective)` can refer to a directive instance for example not a template reference.

aparzi replied:
> The query `ViewChild(MyDirective)` can refer to a directive instance for example not a template reference.

yes I understand, I will try to handle this case. Thanks

aparzi replied:
I updated to handle the case of possible non-string viewChilds, also it should now support multiple viewChilds and viewChildrens

aparzi committed:
fix(core): fix removal of a container reference used in the component file

During migration a container reference was deleted even though it was used in the component file.

thePunderWoman reviewed (approved):
LGTM! Thank you for taking the time to fix this and to add the unit tests.

AndrewKushnir commented:
[Presubmit](http://test/OCL:735545463:BASE:735541757:1741648219748:a6c722c0).

thePunderWoman commented:
Caretaker note: presubmit failures are flakes. This is safe to merge.

AndrewKushnir commented:
This PR was merged into the repository by commit de2bfc09132d19d954d367622c8704993c21da1e.

The changes were merged into the following branches: main, 19.2.x
